### PR TITLE
Fix to make recovered volumes be accounted for by Usage

### PR DESCRIPTION
**Pull request by JoaoJandre**
### Description
When a volume is removed, ACS inserts a `VOLUME.DELETE` record into the `cloud.usage_event` table. If the volume is removed without the expunge option, it can still be restored. 
When the volume is restored, ACS does not insert a new record in the `cloud.usage_event` stating that it has been recreated. Thus, the volume continues to be used without being accounted for.
This PR adds an event in the volume restore process that indicates the volume was recreated. So it will be accounted for.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### How Has This Been Tested?
This PR was tested in a local lab, by destroying and recreating volumes and checking the DB to see if the event was added.
Some unit tests were also added.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@JoaoJandre did you look at the flow for `changeOfferingForVolume`, `migrateVolume`, `resizeVolume` and `updateVolume`?
Not sure but the code you extracted may be used there as well.
Please check, CLGTM otherwise.

**Comment by JoaoJandre** [comment]
> @JoaoJandre did you look at the flow for `changeOfferingForVolume`, `migrateVolume`, `resizeVolume` and `updateVolume`? Not sure but the code you extracted may be used there as well. Please check, CLGTM otherwise.

@DaanHoogland I checked these flows and the event publishes in these places are a bit different, the ones I saw at least already had the `diskOfferingVO` available, so if we used the new method I extracted we would be making an unnecessary call to the DB.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@harikrishna-patnala @GutoVeronezi can you guys have a look?

**Review by rohityadavcloud — APPROVED** [review]

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -1709,9 +1709,30 @@ public Volume recoverVolume(long volumeId) {
...
+        Long offeringId = null;
+        if (diskOfferingId != null) {
+            DiskOfferingVO offering = _diskOfferingDao.findById(diskOfferingId);
+            if (offering != null && !offering.isComputeOnly()) {
+                offeringId = offering.getId();
+            }
+        }
+        UsageEventUtils
+                .publishUsageEvent(EventTypes.EVENT_VOLUME_CREATE, volume.getAccountId(), volume.getDataCenterId(), volume.getId(), volume.getName(), offeringId,
+                        volume.getTemplateId(), volume.getSize(), Volume.class.getName(), volume.getUuid(), volume.isDisplayVolume());
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:1729]:
```suggestion
                        volume.getTemplateId(), volume.getSize(), Volume.class.getName(), volume.getUuid(), volume.isDisplay());
```

As reporter by SonarCloud, `isDisplayVolume` is deprecated; we can use `isDisplay` instead.

**Review by GutoVeronezi — APPROVED** [review]
LGTM

**Comment by DaanHoogland** [comment]
@JoaoJandre almost ready to merge. I might be overcautious but I´d like someone to test the merged code.
@vladimirpetrov can you do that?

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Pull request merged**